File: src/main/java/com/freya02/botcommands/api/waiter/EventWaiter.java
package com.freya02.botcommands.api.waiter;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.internal.utils.EventUtils;
import com.freya02.botcommands.internal.utils.Utils;
import com.freya02.botcommands.internal.waiter.WaitingEvent;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.events.GenericEvent;
import net.dv8tion.jda.api.hooks.EventListener;
import net.dv8tion.jda.api.hooks.SubscribeEvent;
import net.dv8tion.jda.api.requests.GatewayIntent;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.TimeoutException;
import java.util.function.Consumer;
import java.util.function.Predicate;

/**
 * An event waiter - if you need to wait for an event to occur while not blocking threads or having listeners everywhere<br>
 * You provide the type of the JDA event you want to get<br>
 * You can then set properties such as preconditions, timeouts and actions to run when the event gets received / has an exception, etc...
 * <br>This event waiter cannot be constructed and does not need to be registered to the JDA instance, it is already done automatically
 *
 * <h2>Example</h2>
 * <h3>This example uses every actions, has a timeout of 1 second and only triggers if the caller is the same as the user who triggered the previously entered command</h3>
 * <pre><code>
 * final{@literal Future<GuildMessageReceivedEvent>} future = EventWaiter.of(GuildMessageReceivedEvent.class)
 * 		.setOnComplete((f, e, t){@literal ->} System.out.println("Completed"))
 * 		.setOnTimeout((){@literal ->} System.err.println("Timeout"))
 * 		.setOnSuccess(e{@literal ->} System.out.println("Success"))
 * 		.setOnCancelled((){@literal ->} System.err.println("Cancelled"))
 * 		.setTimeout(1, TimeUnit.SECONDS)
 * 		.addPrecondition(e{@literal ->} e.getAuthor().getIdLong() == event.getAuthor().getIdLong())
 * 		.submit();
 *
 * //future.cancel(true);
 *
 * try { //Use this **only** if you need to block your thread waiting for the event, this is considered bad practise so rather use setOnSuccess / setonComplete
 * 	final GuildMessageReceivedEvent guildMessageReceivedEvent = future.get();
 * } catch (InterruptedException | ExecutionException e) {
 * 	e.printStackTrace();
 * }
 * </code></pre>
 */
public class EventWaiter implements EventListener {
	private static final Map<Class<? extends GenericEvent>, List<WaitingEvent<? extends GenericEvent>>> waitingMap = new HashMap<>();
	private static final Logger LOGGER = Logging.getLogger();
	private static final Object EVENT_LIST_LOCK = new Object();

	private static int commandThreadNumber = 0;
	private static final ExecutorService waiterCompleteService = Utils.createCommandPool(r -> {
		final Thread thread = new Thread(r);
		thread.setDaemon(false);
		thread.setUncaughtExceptionHandler((t, e) -> Utils.printExceptionString("An unexpected exception happened in an event waiter thread '" + t.getName() + "':", e));
		thread.setName("Event waiter thread #" + commandThreadNumber++);

		return thread;
	});

	private static JDA jda;
	private static EnumSet<GatewayIntent> intents = EnumSet.noneOf(GatewayIntent.class);
	private static boolean initialized;

	public EventWaiter(@NotNull JDA jda) {
		if (initialized)
			throw new IllegalStateException("Cannot build an EventWaiter more than once");

		initialized = true;

		LOGGER.debug("Initialized EventWaiter");

		EventWaiter.jda = jda;
		EventWaiter.intents = jda.getGatewayIntents();
	}

	/**
	 * Creates a new event waiter builder, waiting for the specified event to occur
	 *
	 * @param eventType The JDA event to wait after
	 * @param <T>       Type of the JDA event
	 * @return A new event waiter builder
	 */
	public static <T extends GenericEvent> EventWaiterBuilder<T> of(Class<T> eventType) {
		if (!initialized)
			throw new IllegalStateException("Framework must be constructed before using the EventWaiter");

		EventUtils.checkEvent(jda, intents, eventType);

		return new EventWaiterBuilder<>(eventType);
	}

	static <T extends GenericEvent> Future<T> submit(WaitingEvent<T> waitingEvent) {
		CompletableFuture<T> future = waitingEvent.getCompletableFuture();

		final List<WaitingEvent<?>> waitingEvents = getWaitingEventsByType(waitingEvent);
		if (waitingEvent.getTimeout() > 0) {
			future.orTimeout(waitingEvent.getTimeout(), waitingEvent.getTimeoutUnit());
		}

		future.whenCompleteAsync((t, throwable) -> {
			final CompletedFutureEvent<T> onComplete = waitingEvent.getOnComplete();
			if (onComplete != null) onComplete.accept(future, t, throwable);

			if (throwable instanceof TimeoutException) {
				final Runnable onTimeout = waitingEvent.getOnTimeout();
				if (onTimeout != null) onTimeout.run();

				synchronized (EVENT_LIST_LOCK) {
					waitingEvents.remove(waitingEvent); //Not removed automatically by Iterator#remove before this method is called
				}
			} else if (t != null) {
				final Consumer<T> onSuccess = waitingEvent.getOnSuccess();
				if (onSuccess != null) onSuccess.accept(t);
			} else if (future.isCancelled()) {
				synchronized (EVENT_LIST_LOCK) {
					waitingEvents.remove(waitingEvent); //Not removed automatically by Iterator#remove before this method is called
				}

				final Runnable onCancelled = waitingEvent.getOnCancelled();
				if (onCancelled != null) onCancelled.run();
			} else {
				LOGGER.warn("Unexpected object received in EventWaiter Future#whenCompleteAsync, please report this to devs");
			}
		}, waiterCompleteService);

		waitingEvents.add(waitingEvent);

		return future;
	}

	@NotNull
	private static <T extends GenericEvent> List<WaitingEvent<?>> getWaitingEventsByType(WaitingEvent<T> waitingEvent) {
		return waitingMap.computeIfAbsent(waitingEvent.getEventType(), x -> Collections.synchronizedList(new ArrayList<>()));
	}

	@SuppressWarnings("unchecked")
	@SubscribeEvent
	@Override
	public void onEvent(@NotNull GenericEvent event) {
		final List<WaitingEvent<? extends GenericEvent>> waitingEvents = waitingMap.get(event.getClass());

		if (waitingEvents != null) {
			synchronized (EVENT_LIST_LOCK) { //Prevent concurrent modification between iterator and on-timeout List#remove
				eventLoop:
				for (Iterator<WaitingEvent<? extends GenericEvent>> iterator = waitingEvents.iterator(); iterator.hasNext(); ) {
					WaitingEvent<? extends GenericEvent> waitingEvent = iterator.next();

					for (Predicate<? extends GenericEvent> p : waitingEvent.getPreconditions()) {
						final Predicate<GenericEvent> precondition = (Predicate<GenericEvent>) p;

						if (!precondition.test(event)) {
							continue eventLoop;
						}
					}

					final CompletableFuture<GenericEvent> completableFuture = (CompletableFuture<GenericEvent>) waitingEvent.getCompletableFuture();
					if (completableFuture.complete(event)) {
						iterator.remove();
					} else {
						LOGGER.warn("Completable future was already completed somehow, please report to the dev");
					}
				}
			}
		}
	}
}

File: BotCommands-core/src/main/kotlin/io/github/freya022/botcommands/api/core/waiter/EventWaiter.kt
package io.github.freya022.botcommands.api.core.waiter

import io.github.freya022.botcommands.api.core.service.annotations.InterfacedService
import kotlinx.coroutines.future.await
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.UserSnowflake
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel
import net.dv8tion.jda.api.events.Event
import net.dv8tion.jda.api.events.message.MessageReceivedEvent

/**
 * Lets you run code when receiving a specific event, while not blocking threads nor having listeners everywhere,
 * and being able to capture existing variables.
 *
 * You can set multiple preconditions, timeouts and actions to run when the event gets received / has an exception, etc...
 *
 * You can also use [JDA.listenOnce] if you wish.
 *
 * ### Example
 * This example uses every action, has a timeout of 1 second and only triggers if the caller is the same as the user who triggered the previously entered command
 *
 * ```java
 * // Get the 'eventWaiter' instance by dependency injection
 * final Future<MessageReceivedEvent> future = eventWaiter.of(MessageReceivedEvent.class)
 *     .setOnComplete((f, evt, throwable) -> System.out.println("Completed with an event " + evt + " or an exception " + throwable))
 *     .setOnTimeout(() -> System.err.println("Timeout"))
 *     .setOnSuccess(evt -> System.out.println("Success, received event: " + evt))
 *     .setOnCancelled(() -> System.err.println("Cancelled"))
 *     .setTimeout(1, TimeUnit.SECONDS)
 *     // Here "event" is the original event, such as a MessageReceivedEvent
 *     .addPrecondition(e -> e.getAuthor().getIdLong() == event.getAuthor().getIdLong())
 *     .submit();
 *
 * // If you want to cancel the event waiter
 * //future.cancel(true);
 *
 * // While I recommend to stick to setOnSuccess / setOnComplete,
 * // you can use this if you need to block while waiting for the event,
 * // but be aware that this might block a thread indefinitely if you're not careful.
 * try {
 *     final MessageReceivedEvent messageReceivedEvent = future.get();
 * } catch (InterruptedException | ExecutionException e) {
 *     LOGGER.error("Error waiting for event", e);
 * }
 * ```
 */
@InterfacedService(acceptMultiple = false)
interface EventWaiter {
    /**
     * Creates a new event waiter builder with the specified event type being awaited.
     *
     * @param eventType The JDA event to wait after
     * @param T         Type of the JDA event
     *
     * @throws IllegalArgumentException
     *         If the event is [RawGatewayEvent][net.dv8tion.jda.api.events.RawGatewayEvent]
     *         but [raw events][net.dv8tion.jda.api.JDABuilder.setRawEventsEnabled] were not enabled.
     *
     * @return A new event waiter builder
     */
    fun <T : Event> of(eventType: Class<T>): EventWaiterBuilder<T>
}

/**
 * Same as [EventWaiter.of] but with a reified type parameter.
 */
inline fun <reified T : Event> EventWaiter.of(): EventWaiterBuilder<T> = of(T::class.java)

/**
 * Suspends until an event of type [T] satisfying the [filter] is received on any shard, then returns it.
 *
 * If you wish to use a timeout with it, you can use [withTimeoutOrNull][kotlinx.coroutines.withTimeoutOrNull].
 *
 * @param filter Condition to satisfy before returning the event
 */
suspend inline fun <reified T : Event> EventWaiter.await(
    noinline filter: (T) -> Boolean = { true },
): T {
    return of<T>()
        .addPrecondition(filter)
        .submit().await()
}

/**
 * Suspends until a [Message] from the [author] satisfying the [filter] is received on any shard, then returns it.
 *
 * If you wish to use a timeout with it, you can use [withTimeoutOrNull][kotlinx.coroutines.withTimeoutOrNull].
 *
 * @param filter Additional conditions to satisfy before returning the message
 */
suspend inline fun EventWaiter.awaitMessage(
    channel: MessageChannel,
    author: UserSnowflake? = null,
    crossinline filter: (Message) -> Boolean = { true }
): Message {
    val channelId = channel.idLong
    val authorId = author?.idLong

    return await<MessageReceivedEvent> {
        it.channel.idLong == channelId
                && (authorId == null || it.author.idLong == authorId)
                && filter(it.message)
    }.message
}
